src/rust/sqs-executor/src/retriever/event_retriever.rs

```diff
@@ -0,0 +1,231 @@
+
+    pub fn get_client(&mut self, region: String) -> S {
+        match self.s3_clients.get(&region) {
+            Some(s3) => s3.clone(),
+            None => {
+                let client = (self.s3_init)(region.clone());
+                self.s3_clients.insert(region.to_string(), client.clone());
+                client
+            }
+        }
+    }
+}
+
+// PayloadDecoder
+
+#[derive(thiserror::Error, Debug)]
+pub enum S3PayloadRetrieverError<DecoderErrorT>
+where
+    DecoderErrorT: CheckedError + 'static,
+{
+    #[error("S3Error: {0}")]
+    S3Error(#[from] RusotoError<GetObjectError>),
+    #[error("Decode error")]
+    DecodeError(#[from] DecoderErrorT),
+    #[error("IO")]
+    Io(#[from] std::io::Error),
+    #[error("JSON")]
+    Json(#[from] serde_json::Error),
+    #[error("Timeout")]
+    Timeout(#[from] Elapsed),
+}
+
+impl<DecoderErrorT> CheckedError for S3PayloadRetrieverError<DecoderErrorT>
+where
+    DecoderErrorT: CheckedError + 'static,
+{
+    fn error_type(&self) -> Recoverable {
+        match self {
+            Self::S3Error(_) => Recoverable::Transient,
+            Self::DecodeError(_) => Recoverable::Persistent,
+            Self::Io(_) => Recoverable::Transient,
+            Self::Json(_) => Recoverable::Persistent,
+            Self::Timeout(_) => Recoverable::Transient,
+        }
+    }
+}
+
+#[async_trait]
+impl<S, SInit, D, E, DecoderErrorT> PayloadRetriever<E>
+    for S3PayloadRetriever<S, SInit, D, E, DecoderErrorT>
+where
+    S: S3 + Clone + Send + Sync + 'static,
+    SInit: (Fn(String) -> S) + Clone + Send + Sync + 'static,
+    D: PayloadDecoder<E, DecoderError = DecoderErrorT> + Clone + Send + 'static,
+    E: Send + 'static,
+    DecoderErrorT: CheckedError + Send + 'static,
+{
+    type Message = SqsMessage;
+    type Error = S3PayloadRetrieverError<DecoderErrorT>;
+
+    #[tracing::instrument(skip(self, msg))]
+    async fn retrieve_event(&mut self, msg: &Self::Message) -> Result<Option<E>, Self::Error> {
+        let body = msg.body.as_ref().unwrap();
+        println!("Got body from message: {}", body);
+        debug!("Got body from message: {}", body);
+        let event: serde_json::Value = serde_json::from_str(body)?;
+
+        if let Some(Some(event_str)) = event.get("Event").map(serde_json::Value::as_str) {
+            if event_str == "s3:TestEvent" {
+                return Ok(None);
+            }
+        }
+        let record = &event["Records"][0]["s3"];
+
+        let bucket = record["bucket"]["name"].as_str().expect("bucket name");
+        let key = record["object"]["key"].as_str().expect("object key");
+        debug!("Retrieving S3 payload from: {} / {}", bucket, key);
+        let region = &event["Records"][0]["awsRegion"].as_str().expect("region");
+        let s3 = self.get_client(region.to_string());
+        let s3_data = s3.get_object(GetObjectRequest {
+            bucket: bucket.to_string(),
+            key: key.to_string(),
+            ..Default::default()
```

Comment:
(I've done 2 minutes of rust doc perusing and couldn't find this syntax; is it a destructure?)